duet/gps.ml

```diff
@@ -0,0 +1,747 @@
+open Core
+open Srk
+open CfgIr
+open BatPervasives
+open Cra
+
+module TS = TransitionSystem.Make(Ctx)(V)(K)
+
+include Log.Make(struct let name = "gps" end)
+
+module ProcName = struct
+  type t = int * int
+
+  let make ((u, v) : TS.vertex * TS.vertex) : t = (u, v)
+
+  let string_of (p: t) =
+    let u, v = p in Printf.sprintf "%d:%d" u v
+
+  let of_string (s: string) =
+    match String.split_on_char ':' s with
+    | [ us ; vs ] -> (make ((int_of_string us), (int_of_string vs)))
+    | _ -> failwith @@ Printf.sprintf "illegal procedure identifier %s" s
+
+  (* lexicographic comparison using Stdlib.compare *)
+  let compare (p1: t) (p2: t) = Stdlib.compare p1 p2
+  let hash = Hashtbl.hash
+  let equal = (=)
+end
+
+module ProcHT = BatHashtbl.Make(ProcName)
+module ProcMap = BatMap.Make(ProcName)
+module IntMap = BatMap.Make(Int)
+module StringMap = BatMap.Make(String)
+module DQ = BatDeque
+module ARR = Batteries.DynArray
+type cfg_t = TSG.t
+type idq_t = int BatDeque.t
+type state_formula = Ctx.t Syntax.formula
+exception Mexception of string
+let mk_true () = Syntax.mk_true Ctx.context
+let mk_false () = Syntax.mk_false Ctx.context
+let mk_query ts entry = TS.mk_query ts entry (if !monotone then (module MonotoneDom) else (module TransitionDom))
+
+let log_formulas prefix formulas =
+  List.iteri (fun i f -> logf "[formula] %s(%i): %a\n" prefix i (Syntax.pp_expr srk) f) formulas
+
+let log_weights prefix weights =
+  List.iteri (fun i f -> logf "[weight] %s(%i): %a\n" prefix i K.pp f) weights
+
+
+let log_model prefix model =
+  logf "[model] %s: %a\n" prefix Interpretation.pp model
+
+(*
+let assert_i = ref 0
+let new_assert_var cond =
+  let i = !assert_i in
+  let name = "__assert" ^ (string_of_int i) in
+  let v = Varinfo.mk_global name (Concrete (Int 8)) |> Var.mk in
+  let assert_var = Syntax.mk_symbol srk ~name:name `TyInt in
+  let assert_term = Syntax.mk_const srk assert_var in
+  assert_i := !assert_i + 1;
+  K.assign v cond
+
+let process_interproc_assertion (ts: cfg_t) (phi: Ctx.formula) v =
+  let a_var, a_term = new_assert_var @@ Ctx.mk_not phi in
+
+*)
+
+(* Convert assertion checking problem to vertex reachability problem. *)
+let safety_to_reachability (ts : cfg_t) assertions =
+  let err_loc = 1 + (WG.fold_vertex (fun v max -> if v > max then v else max) ts 0) in
+  let ts = WG.add_vertex ts err_loc in
+  let ts =
+    SrkUtil.Int.Map.fold (fun v (phi, _, _) ts ->
+        let s = Printf.sprintf " Adding assertion node %d -> %d for label " v err_loc in
+        log_formulas s [ Ctx.mk_not phi ] ;
+        WG.add_edge ts v (Weight (K.assume (Ctx.mk_not phi))) err_loc)
+      assertions
+      ts
+  in
+  (ts, err_loc)
+
+module Summarizer =
```

Comment:
I think we can omit all of the interprocedural logic until we have a better idea of what it's going to look like in the end.